internal/server/websocketServer.go

```diff
@@ -0,0 +1,185 @@
+	"sync"
+	"time"
+
+	"github.com/dicedb/dice/config"
+	"github.com/dicedb/dice/internal/clientio"
+	"github.com/dicedb/dice/internal/ops"
+	"github.com/dicedb/dice/internal/querywatcher"
+	"github.com/dicedb/dice/internal/server/utils"
+	"github.com/dicedb/dice/internal/shard"
+	dstore "github.com/dicedb/dice/internal/store"
+	"github.com/gorilla/websocket"
+)
+
+var unimplementedCommandsWebsocket map[string]bool = map[string]bool{
+	"QWATCH":    true,
+	"QUNWATCH":  true,
+	"SUBSCRIBE": true,
+	Abort:       false,
+}
+
+type WebsocketServer struct {
+	querywatcher    *querywatcher.QueryManager
+	shardManager    *shard.ShardManager
+	ioChan          chan *ops.StoreResponse
+	watchChan       chan dstore.WatchEvent
+	websocketServer *http.Server
+	upgrader        websocket.Upgrader
+	logger          *slog.Logger
+	shutdownChan    chan struct{}
+}
+
+func NewWebSocketServer(shardManager *shard.ShardManager, watchChan chan dstore.WatchEvent, logger *slog.Logger) *WebsocketServer {
+	mux := http.NewServeMux()
+	srv := &http.Server{
+		Addr:              fmt.Sprintf(":%d", config.WebsocketPort),
+		Handler:           mux,
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+
+	websocketServer := &WebsocketServer{
+		shardManager:    shardManager,
+		querywatcher:    querywatcher.NewQueryManager(logger),
+		ioChan:          make(chan *ops.StoreResponse, 1000),
+		watchChan:       watchChan,
+		websocketServer: srv,
+		upgrader:        upgrader,
+		logger:          logger,
+		shutdownChan:    make(chan struct{}),
+	}
+
+	mux.HandleFunc("/", websocketServer.WebsocketHandler)
+	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte("ok"))
+		if err != nil {
+			return
+		}
+	})
+	return websocketServer
+}
+
+func (s *WebsocketServer) Run(ctx context.Context) error {
+	var wg sync.WaitGroup
+	var err error
+
+	websocketCtx, cancelWebsocket := context.WithCancel(ctx)
+	defer cancelWebsocket()
+
+	s.shardManager.RegisterWorker("wsServer", s.ioChan)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-s.shutdownChan:
+			err = ErrAborted
+			s.logger.Debug("Shutting down Websocket Server")
+		}
+
+		shutdownErr := s.websocketServer.Shutdown(websocketCtx)
+		if shutdownErr != nil {
+			s.logger.Error("Websocket Server shutdown failed:", slog.Any("error", err))
+			return
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		s.logger.Info("Websocket Server running", slog.String("port", s.websocketServer.Addr[1:]))
+		err = s.websocketServer.ListenAndServe()
+	}()
+
+	wg.Wait()
+	return err
+}
+
+func (s *WebsocketServer) WebsocketHandler(w http.ResponseWriter, r *http.Request) {
+	// upgrade http connection to websocket
+	conn, err := s.upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		s.logger.Error("Websocket upgrade failed", slog.Any("error", err))
+	}
+	defer conn.Close()
```

Comment:
We should do a normal closure. Check for 1000 error code.
We can have closing handshake implemented and not abruptly close the WS connection.

Reply:
Added.